rust-examples/src/patternmatching.rs

```diff
@@ -0,0 +1,36 @@
+enum QuadraticSolution {
+    NoSolution,
+    OneSolition(f64),
+    TwoSolutions(f64, f64),
+}
+
+fn solve_quadratic_equation(a: f64, b: f64, c: f64) -> QuadraticSolution {
+    let discriminant = b * b - 4.0 * a * c;
+    if discriminant < 0.0 {
+        QuadraticSolution::NoSolution
+    } else if discriminant == 0.0 {
+        QuadraticSolution::OneSolition(-b / (2.0 * a))
+    } else {
+        QuadraticSolution::TwoSolutions(
+            (-b + discriminant.sqrt()) / (2.0 * a),
+            (-b - discriminant.sqrt()) / (2.0 * a),
+        )
+    }
+}
+
+fn main() {
+    let res = solve_quadratic_equation(1.0, 3.0, 1.0);
+    match res {
+        QuadraticSolution::NoSolution => println!("No solution"),
+        QuadraticSolution::OneSolition(x) => println!("One solution: {}", x),
+        QuadraticSolution::TwoSolutions(x, y) => println!("Two solutions: {}, {}", x, y),
```

Comment:
so, personally I'd focus more on the power of patternmatching here– right now the example mainly looks like a switch statement without having to type breaks.
To me, what makes match statements so powerful is:
- using `_` to explicitly ignore values
- having access to [match guards](https://doc.rust-lang.org/reference/expressions/match-expr.html#match-guards)
- having the compiler shout at me when I forget to cover a case
- seeing patterns is more intuitive to me visually/personally?

So, personally, I'd start with a simple *and incomplete* if...else and then have participants
1. think about how they can turn it into a match statement
2. figure out how to make the compiler shut up about cases they don't care about/are grouped

and then compare what they came up with & maybe discuss different ways of approaching match statements.

Reply:
so, the starting point could look something like this: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=5d6ac066c7e7efb03e46cca8cf81ab20

Reply:
and then one possible solution is: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=5f7687e53432df693ec93e9cdf7ad244

Reply:
sorry for being so reactive about this :( I think it's important to keep confusion overhead low because people can only digest so many things at once, and so I'd go with examples that are as simple as possible, even though there are so many additional clever things one wants to show off

Reply:
> and then one possible solution is: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=5f7687e53432df693ec93e9cdf7ad244

this could then be extended in the demo to show match guards, having boolean expressions (for example by introducing `FarmAnimal::Bull` and then matching `FarmAnimal::Cow | FarmAnimal::Bull => Some("moo".to_string())`)

Reply:
I think we should definitely use your examples as an introduction to pattern matching and maybe mine as an example to illustrate how something very boilerplate-y like std::variant and std::visit can be written more easily in Rust.

Reply:
oh yes that sounds like a great idea! (looking forward to learning a lot aout C++ on wednesday :D) So you want me to open a separate PR with those intro examples?

Reply:
If you have the time you can put your three examples in a separate PR. If you dont have the time I can do that as well tomorrow.

Reply:
done https://github.com/hniemeyer/RustForCppDevsCodeSnippets/pull/8